cirq/protocols/commutes_protocol.py

```diff
@@ -0,0 +1,223 @@
+
+    Raises:
+        TypeError: default was not set and there was no implemented rule for
+            efficiently determining if the two values commute.
+    """
+    atol = float(atol)
+
+    strats = [
+        _strat_commutes_from_commutes,
+        _strat_commutes_from_matrix,
+        _strat_commutes_from_operation,
+    ]
+    for strat in strats:
+        result = strat(v1, v2, atol=atol)
+        if result is None:
+            break
+        if result is not NotImplemented:
+            return result
+    if default is not RaiseTypeErrorIfNotProvided:
+        return default
+    raise TypeError(
+        f"Failed to determine whether or not "
+        f"{v1!r} commutes with {v2!r}. "
+        f"The result may be indeterminate, or there may be no strategy "
+        f"implemented to handle this case.\n"
+        f"If you want a default result in this case, specify a `default=` "
+        f"argument or use `cirq.definitely_commutes`.")
+
+
+def definitely_commutes(v1: Any, v2: Any, *,
+                        atol: Union[int, float] = 1e-8) -> bool:
+    """Determines whether two values definitely commute.
+
+    Returns:
+        True: The two values definitely commute.
+        False: The two values may or may not commute.
+    """
+    return commutes(v1, v2, atol=atol, default=False)
+
+
+def _strat_commutes_from_commutes(v1: Any,
+                                  v2: Any,
+                                  *,
+                                  atol: Union[int, float] = 1e-8
+                                 ) -> Union[bool, NotImplementedType, None]:
+    """Attempts to determine commutativity via the objects' _commutes_
+    method."""
+
+    for a, b in [(v1, v2), (v2, v1)]:
+        getter = getattr(a, '_commutes_', None)
+        if getter is None:
+            continue
+        val = getter(b, atol=atol)
+        if val is not NotImplemented:
+            return val
+    return NotImplemented
+
+
+def _strat_commutes_from_matrix(
+        v1: Any,
+        v2: Any,
+        *,
+        atol: float,
+) -> Union[bool, NotImplementedType, None]:
+    """Attempts to determine commutativity of matrices."""
+    if not isinstance(v1, np.ndarray) or not isinstance(v2, np.ndarray):
+        return NotImplemented
+    if v1.shape != v2.shape:
+        return None
+    return linalg.matrix_commutes(v1, v2, atol=atol)
+
+
+def _strat_commutes_from_operation(
+        v1: Any,
+        v2: Any,
+        *,
+        atol: float,
+) -> Union[bool, NotImplementedType, None]:
+    if not isinstance(v1, ops.Operation) or not isinstance(v2, ops.Operation):
+        return NotImplemented
+
+    if set(v1.qubits).isdisjoint(v2.qubits):
+        return True
+
+    from cirq import circuits
+    circuit12 = circuits.Circuit(v1, v2)
+    circuit21 = circuits.Circuit(v2, v1)
+
+    # Don't create gigantic matrices.
+    if np.product(qid_shape_protocol.qid_shape(circuit12)) > 2**10:
+        return NotImplemented  # coverage: ignore
```

Comment:
This should be easy enough to cover.

Reply:
The issue is that I don't want to write a test asserting that something doesn't work, because in the future we may decide that it should work. That would be pinning down implementation details instead of ensuring stuff works.